source/opt/constant_value.h

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef LIBSPIRV_OPT_CONSTANT_VALUE_H_
+#define LIBSPIRV_OPT_CONSTANT_VALUE_H_
+
+#include <memory>
+#include <vector>
+
+#include "instruction.h"
+
+namespace spvtools {
+namespace opt {
+
+// Class hierarchy to record the in-operand of front-end constants.
+// ScalarValue stores the value of of scalar constants. CompositeValue stores a
+// vector of its component ids. Value is a abstract class of ScalarValue and
+// CompositeValue.
+
+class ScalarConst;
+class BoolConst;
+class CompositeConst;
+class VectorConst;
+class NullConst;
+
+class Const {
+ public:
+  Const(uint32_t result_id, uint32_t type_id)
+      : id_(result_id), type_id_(type_id) {}
+
+  virtual std::unique_ptr<Const> MakeCopyWithId(uint32_t result_id) const = 0;
```

Comment:
Or could have named it CloneWithId which is shorter and probably just as descriptive.
Coder's choice.